crates/compiler/can/src/suffixed.rs

```diff
@@ -813,33 +836,122 @@ fn unwrap_low_level_dbg<'a>(
     }
 }
 
-/// Helper for `Task.await (loc_arg) \loc_pat -> loc_new`
+/// Helper for `Task.await loc_expr \loc_pat -> loc_cont`
 pub fn apply_task_await<'a>(
     arena: &'a Bump,
     region: Region,
-    loc_arg: &'a Loc<Expr<'a>>,
+    loc_expr: &'a Loc<Expr<'a>>,
     loc_pat: &'a Loc<Pattern<'a>>,
-    loc_new: &'a Loc<Expr<'a>>,
+    loc_cont: &'a Loc<Expr<'a>>,
+    maybe_loc_ann: Option<(&'a Loc<Pattern>, &'a Loc<TypeAnnotation<'a>>)>,
 ) -> &'a Loc<Expr<'a>> {
-    // If the pattern and the new are matching answers then we don't need to unwrap anything
-    // e.g. `Task.await foo \#!a1 -> Task.ok #!a1` is the same as `foo`
-    if is_matching_intermediate_answer(loc_pat, loc_new) {
-        return loc_arg;
-    }
+    let task_await_first_arg = match maybe_loc_ann {
+        Some((loc_ann_pat, loc_type)) => {
+            // loc_ann_pat : loc_type
+            // loc_pat = loc_expr!
+            // loc_cont
+
+            // desugar to
+            // Task.await
+            //     (
+            //         #!0_expr : Task loc_type _
+            //         #!0_expr = loc_expr
+            //         #!0_expr
+            //     )
+            //     \loc_pat -> loc_cont
+            use roc_parse::ast::*;
+
+            // #!0_expr or #!0_stmt
+            let new_ident = next_unique_suffixed_ident();
+            let new_ident = match loc_pat.value {
+                Pattern::Underscore("#!stmt") => format!("{}_stmt", new_ident),
+                Pattern::Identifier { ident }
+                    if ident.starts_with('#') && ident.ends_with("_stmt") =>
```

Comment:
This seems a little fishy; why do we need to do string parsing/matching like this? This is mildly concerning since it introduces a very subtle dependency, where someone changing a string that shouldn't matter, can accidentally have broader consequences in the future.

Is there a way to get this same information without trying to parse identifier strings like this?

Reply:
Sure. Need to change the existing enum structure or introduce another one. I just didn't want to have too many changes in this pr to keep it simple. I can refactor it as a follow-up